File: engine/src/schemas/risk.py
"""Pydantic models for Risk Profile entity."""

from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, Field


class RiskModificationLogEntry(BaseModel):
    """Audit log entry for risk profile changes."""
    timestamp: datetime
    field: str
    previous_value: Any
    new_value: Any
    confirmed_by_session: str


class RiskProfile(BaseModel):
    """Risk profile with non-bypassable enforcement."""
    id: UUID
    name: str
    max_drawdown_pct: float = Field(ge=0.01, le=1.0)
    single_order_size_cap: float = Field(gt=0)
    daily_loss_limit: float = Field(gt=0)
    max_concurrent_strategies: int = Field(ge=1)
    kill_switch_active: bool = False
    created_at: datetime
    updated_at: datetime
    modification_log: list[RiskModificationLogEntry] | None = None


class RiskProfileUpdate(BaseModel):
    """Request body for updating risk profile."""
    max_drawdown_pct: float | None = Field(default=None, ge=0.01, le=1.0)
    single_order_size_cap: float | None = Field(default=None, gt=0)
    daily_loss_limit: float | None = Field(default=None, gt=0)
    max_concurrent_strategies: int | None = Field(default=None, ge=1)
    confirmed_by_session: str

File: packages/shared/src/schemas/risk.ts
import { z } from 'zod';

export const RiskModificationLogEntrySchema = z.object({
  timestamp: z.string().datetime(),
  field: z.string(),
  previous_value: z.unknown(),
  new_value: z.unknown(),
  confirmed_by_session: z.string(),
});

export const RiskProfileSchema = z.object({
  id: z.string().uuid(),
  name: z.string(),
  max_drawdown_pct: z.number().min(0.01).max(1.0),
  single_order_size_cap: z.number().positive(),
  daily_loss_limit: z.number().positive(),
  max_concurrent_strategies: z.number().int().min(1),
  kill_switch_active: z.boolean(),
  created_at: z.string().datetime(),
  updated_at: z.string().datetime(),
  modification_log: z.array(RiskModificationLogEntrySchema).optional(),
});

export const RiskProfileUpdateSchema = z.object({
  max_drawdown_pct: z.number().min(0.01).max(1.0).optional(),
  single_order_size_cap: z.number().positive().optional(),
  daily_loss_limit: z.number().positive().optional(),
  max_concurrent_strategies: z.number().int().min(1).optional(),
  confirmed_by_session: z.string(),
});

export type RiskProfile = z.infer<typeof RiskProfileSchema>;
export type RiskProfileUpdate = z.infer<typeof RiskProfileUpdateSchema>;
export type RiskModificationLogEntry = z.infer<typeof RiskModificationLogEntrySchema>;
